qcore/geo.py

```diff
@@ -827,3 +830,300 @@ def compute_intermediate_latitudes(lon_lat1, lon_lat2, lon_in):
             / (np.cos(lat1) * np.cos(lat2) * np.sin(lon1 - lon2))
         )
     ) / conversion_factor
+
+
+def homogenise_point(p: np.ndarray) -> np.ndarray:
+    """Express a point in homogenous coordinates.
+
+    Given a point p in R^3 return the affine point in PG(3, R) associated with p. Informally, it maps
+    (x, y, z) -> (x, y, z, 1).
+
+    Parameters
+    ----------
+    p : np.ndarray
+        A point in R^3.
+
+    Returns
+    -------
+    np.ndarray
+        The associated point in PG(3, R).
+
+    Examples
+    --------
+    >>> homogenise_point(np.array([1, 0, 0]))
+    array([1,0,0,1])
+    """
+
+    return np.concatenate((p, np.array([1])))
+
+
+def projective_span(p: np.ndarray, q: np.ndarray, r: np.ndarray) -> np.ndarray:
+    """Find the projective plane spanned by p, q, and r.
+
+    Parameters
+    ----------
+    p : np.ndarray
+        The homogenous coordinates of p.
+    q : np.ndarray
+        The homogenous coordinates of q.
+    r : np.ndarray
+        The homogenous coordinates of r.
+
+    Returns
+    -------
+    np.ndarray
+        The dual coordinates of the plane spanning p, q, and r
+
+    Raises
+    ------
+    ValueError
+        If the supplied points don't span a plane in PG(3, R)
+
+    Examples
+    --------
+    >>> projective_span(np.array([1, 0, 0, 1]), np.array([1, 1, 0, 1]), np.array([0, 0, 0, 1]))
+    np.array([0, 0, 1, 0])
+    >>> projective_span(np.array([0, 0, 0, 1]), np.array([1, 1, 0, 1]), np.array([0.5, 0.5, 0, 1]))
+    ValueError: Points supplied do not span a plane. # NOTE: these points lie on the line x = y.
+    """
+    M = np.vstack([p, q, r])
+    ns = sp.linalg.null_space(M)
+    # If the null space does not have dimension 1 then the points supplied do
+    # not span a unique plane. This could happen if the points supplied are
+    # collinear.
+    if ns.shape[1] != 1:
+        raise ValueError("Points supplied do not span a plane.")
+    # Otherwise we just rescale the homogenous vector. It is now mathematically
+    # impossible that the vector is zero, so we can safely assume that c != 0 in the following code.
+    # We normalise by the last non-zero coordinate of the point.
+    ns = ns.reshape((1, -1))[0]
+    c = next(x for x in reversed(ns) if not np.isclose(x, 0))
+    return ns / c
```

Comment:
Maybe some easier to understand variable names other than p, q, r, M, ns, c and x could be used? Unless there is a mathematical definition reason for doing so?

Reply:
I found this part difficult to follow.  ns is supposed to have dimension 1 to span a unique plane, and we check ns.shape[1] != 1..? Could you also explain why "dividing by the last non-zero coordinate" makes it normalized?

Reply:
Right, it's better to think of the null space not as the span of p, q, and r, but as the space of all planes with dual coordinates x such that x * p == 0, x * q == 0, x * r == 0. In this context, if you have three distinct non-collinear points, they are contained within a unique plane with some dual coordinates x. However the dual coordinates for this plane are not unique. In particular, any scalar multiple of x will define the same plane.  Thus, the null space will have dimension 1 in this case (to represent all scalar multiples of x). 

The normalisation step is simply to pick one "good" representation for x. As discussed, while the plane spanning p, q, and r is unique, the dual coordinates for x are not necessarily unique. In particular, I have chosen to scale x such that the last non-zero coordinate is one (which is achieved by dividing through by that coordinate). This choice is arbitrary, and shares the same name as normalising by dividing through by the length of x it's entirely unrelated.

For an example of how this works, suppose we are trying to find the plane containing the points (0, 0, 0), (0, 1, 0) and (0, 0, 1). We first add the extra 1 coordinate:

(0, 0, 0, 1)
(0, 1, 0, 1)
(0, 0, 1, 1)

and then we compute the null space, which is one-dimensional and has a single basis vector (1, 0, 0, 0). This represents the plane with equation x = 0. Any scalar multiple of these coordinates would suffice as the dual coordinates, (2, 0, 0, 0) represents the equation 2x = 0 (which is the same plane). But because we need to make one choice and stick to it, the code will rescale the solution to get (1, 0, 0, 0).

I'll shorten this and change the description to make it clearer what is going on.

Reply:
Oh yeah in the cases where the dimension of the null space is not 1, there are multiple inequivalent dual coordinates x that satisfy x * p, x * q, x * r == 0. e.g.

(1, 0, 0, 1)
(0, 1, 0, 1)
(1/2, 1/2, 0, 1)

These points are collinear, and the null space of that matrix has rank 2 and is spanned by the vectors (0, 0, 1, 0) and (1, 1, 0, -1). This is to represent the fact that there are multiple different planes that contain these points, e.g. y = 0 and 1/2 x + 1/2 y - 1 = 0.